### ISSUE TITLE
NotCamelCaps complains about API from php extension

### ISSUE THREAD

**Issue by greg0ire** [squizlabs/PHP_CodeSniffer#2950]
[I'm trying to update phpcs on `doctrine/dbal`](https://github.com/doctrine/dbal/pull/3983).
We have a piece of code where we do the following:

```php
        $serverInfo = db2_server_info($this->conn);

        return $serverInfo->DBMS_VER;
```

phpcs produces this error:

```
FILE: lib/Doctrine/DBAL/Driver/IBMDB2/DB2Connection.php
-----------------------------------------------------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------------------------------------
 67 | ERROR | Variable "DBMS_VER" is not in valid camel caps format
    |       | (Squiz.NamingConventions.ValidVariableName.NotCamelCaps)
-----------------------------------------------------------------------------------------------------------------------------
```

Since we have no control over this external API, I would not expect an error.

**Comment by gsherwood** [comment]
There is no way for the sniff to know that this isn't your variable. All you can do is mute the error:
```
// phpcs:ignore Squiz.NamingConventions.ValidVariableName
return $serverInfo->DBMS_VER;
```

**Comment by greg0ire** [comment]
Oh ok… I'll close then, cc @morozov

**Comment by morozov** [comment]
> There is no way for the sniff to know that this isn't your variable.

Does this mean that the sniff cannot tell whether it's property access or property declaration?

**Comment by gsherwood** [comment]
> Does this mean that the sniff cannot tell whether it's property access or property declaration?

It knows how it is accessed, but it reports on all usages and you can't differentiate between them.

**Comment by morozov** [comment]
Would it make sense to not report the usage? There's no point in reporting an issue that cannot be fixed by design. If the declaration can be fixed, the usage will no longer violate the rule.

**Comment by gsherwood** [comment]
> Would it make sense to not report the usage?

A better change would probably be to adjust the error codes so developers could decide to ignore the various usages. Both types of changes would be BC breaks though.

**Comment by morozov** [comment]
Could it be defined via the sniff configuration/properties instead? E.g. `$reportUsages = true` by default that could be disabled in a specific ruleset?

**Comment by gsherwood** [comment]
> Could it be defined via the sniff configuration/properties instead?

It could, but I prefer using sniff codes. The sniff properties are an older way of achieving the same thing.

**Comment by morozov** [comment]
Hm… with properties, you can define as many of them it appears to be needed without breaking the existing usages while every time you introduce a new code, you'll introduce a break as you said above.

In any event, if you agree that this is a legit issue worth fixing, could you reopen it?

**Comment by gsherwood** [comment]
> In any event, if you agree that this is a legit issue worth fixing, could you reopen it?

I have no intention of working on this, so I'll keep it closed. If someone wants to work on a PR, I'd consider it, but only in version 4 due to the BC break.

**Comment by kamazee** [comment]
> If someone wants to work on a PR, I'd consider it, but only in version 4 due to the BC break

If reporting of usages is on by default, behavior remains the same with an option for those who want it to differ: if someone is willing to address the issue this way, will such PR be considered?

**Comment by gsherwood** [comment]
> If reporting of usages is on by default, behavior remains the same with an option for those who want it to differ

The reason I suggested version 4 is because I'd prefer the sniff to use the error code to indicate how the variable was used. This allows developers to change the message and severity (or ignore the error entirely) by using that error code.

Sniff properties are an older way of achieving a similar thing, but they provide less flexibility as you need a setting for each type of usage, and then the errors are presented with the same message and code.

In this case, you may need an `ignoreMemberVars` setting, but then developers can't decide to still flag these as warnings for manual review. So you then add a `memberVarsSeverity` setting, and you've just replicated 2 features of the phpcs.xml format.

Hope that makes sense.
